service/start.go

```diff
@@ -14,67 +15,66 @@ import (
 )
 
 // Start a service
-func (service *Service) Start() (serviceIDs []string, err error) {
+func (service *Service) Start() ([]string, error) {
 	status, err := service.Status()
 	if err != nil || status == RUNNING {
-		return //TODO: if the service is already running, serviceIDs should be returned.
+		return nil, err //TODO: if the service is already running, serviceIDs should be returned.
 	}
 	// If there is one but not all services running stop to restart all
 	if status == PARTIAL {
-		err = service.StopDependencies()
-		if err != nil {
-			return
+		if err = service.StopDependencies(); err != nil {
+			return nil, err
 		}
 	}
 	networkID, err := container.CreateNetwork(service.namespace())
 	if err != nil {
-		return
+		return nil, err
 	}
-	dependenciesFromService := service.DependenciesFromService()
-	serviceIDs = make([]string, len(dependenciesFromService))
-	var mutex sync.Mutex
-	var wg sync.WaitGroup
+	var (
+		mutex                   sync.Mutex
+		wg                      sync.WaitGroup
+		dependenciesFromService = service.DependenciesFromService()
+		serviceIDs              = make([]string, len(dependenciesFromService))
+	)
 	for i, dependency := range dependenciesFromService {
 		wg.Add(1)
 		go func(dep *DependencyFromService, i int) {
-			defer wg.Done()
 			serviceID, errStart := dep.Start(networkID)
 			mutex.Lock()
-			defer mutex.Unlock()
 			serviceIDs[i] = serviceID
 			if errStart != nil && err == nil {
 				err = errStart
 			}
+			mutex.Unlock()
```

Comment:
I'm curious on this one, why removing the `defer` the end result will be the same because anyway we are doing it in the end of the function. Is it for optimization or is their any other reasons ?